docs: translate content/docs/thinking-in-react.md and other related (#19)
reactjs/tl.reactjs.org: content/docs/thinking-in-react.md

---
id: thinking-in-react
title: Paano mag-isip sa React
permalink: docs/thinking-in-react.html
redirect_from:
  - 'blog/2013/11/05/thinking-in-react.html'
  - 'docs/thinking-in-react-zh-CN.html'
prev: composition-vs-inheritance.html
---

> Ang React ay, sa aming opinyon, nangunguna sa pagbuo ng malaki at mabilis na Web apps gamit ang Javascript. Ito ay mabisa naming nagamit sa Facebook at Instagram.

Reviewer: ```suggestion
Ang React, sa aming opinyon, ay nangunguna pagdating sa pagbuo ng malaki at mabilis na Web apps gamit ang Javascript na mabisa naming nagamit sa Facebook at Instagram.
```

> Isa sa mga malaking parte ng React ay kung paano ka mapapaisip tungkol sa mga apps habang binubuo mo ito. Sa dokyumentong ito, ipapakita namin sa inyo ang thought process sa pagbuo ng searchable product data table gamit ang React.

Reviewer: ```suggestion
Isa sa malaking parte ng React ay kung paano ka magisip habang bumubuo ka ng apps. Sa dokyumentong ito, ipapakita namin sa inyo ang thought process sa pagbuo ng searchable product data table gamit ang React.
```

> ## Umpisahan gamit ang Mock {#start-with-a-mock}

Reviewer: ```suggestion
## Magumpisa gamit ang isang Mock {#start-with-a-mock}
```

];
```

## Ika-1 Hakbang: Hatiin ang UI sa isang Component Hierarchy {#step-1-break-the-ui-into-a-component-hierarchy}

> Ang una mong dapat gawin ay gumuhit ng mga kahon sa paligid ng bawat component (at subcomponent) sa mock at bigyan mo lahat ito ng pangalan. Kung ikaw ay may katuwang na designer, maaring sila mismo ay nagawa na ito, bilis at makipag-ugnayan sa kanila! Ang kanilang Photoshop layer names ay maaring pangalan ng iyong React components.

Reviewer: ```suggestion
Ang una mong dapat gawin ay gumuhit ng mga kahon sa paligid ng bawat component (at subcomponent) sa mock at bigyan mo lahat ito ng pangalan. Kung ikaw ay may katuwang na designer, maaring nagawa na ito, at maaari kang makipag-ugnayan sa kanila. Maaari mo ding pangalanan ang iyong React components gamit ang kanilang Photoshop layer name.
```

> Ngunit paano mo malalaman kung ano ba dapat ang iyong mga component? Gamiting ang parehong pamamaraan sa pagpasya kung ikaw ay gagawa ng bagong function o object. Isang pamamaraan ay ang [single responsibility principle](https://en.wikipedia.org/wiki/Single_responsibility_principle), ito ay, kung saan ang component ay dapat isa lang ang tungkulin. Kung ang component ay lumalaki, dapat na mahati ito sa mas maliliit pang subcomponents.

Reviewer: ```suggestion
Ngunit paano mo malalaman kung anong mga component ang iyong gagawin? Gamiting ang iyong pamamaraan sa pagpasya bago ka gumawa ng bagong function o object. O kaya gamitin ang [single responsibility principle](https://en.wikipedia.org/wiki/Single_responsibility_principle) na nagsasabing ang isang component ay dapat isa lang ang tungkulin. Kung ang component ay lumalaki, dapat na mahati ito sa mas maliliit pang subcomponents.
```

> Dahil madalas kang nagpapakita ng JSON data model sa user, mapapasin mo na kung ang iyong model ay tama ang pagkakagawa, ang iyong UI (at pati ang mga component structure) ay sasangayon ng maayos. Sa kadahilanang ang UI at data models ay sumusunod sa kaparehong *information architecture*. Hatiin ang iyong UI sa mga components, kung saan ang bawat compoent ay tugma sa iyong data model.

Reviewer: ```suggestion
Dahil madalas kang nagpapakita ng JSON data model sa user, mapapasin mo kung tama ang pagkakagawa ng iyong model, ang iyong UI (at pati ang iyong component structure) ay sasangayon ng maayos. Sa kadahilanang ang UI at data models ay sumusunod sa kaparehong *information architecture*. Hatiin ang iyong UI sa mga components, kung saan ang bawat component ay tugma sa iyong data model.
```

  * `FilterableProductTable`
    * `SearchBar`
    * `ProductTable`
      * `ProductCategoryRow`
      * `ProductRow`

## Ika-2 Hakbang: Gumawa ng Static Version sa React {#step-2-build-a-static-version-in-react}

<p data-height="600" data-theme-id="0" data-slug-hash="BwWzwm" data-default-tab="js" data-user="lacker" data-embed-version="2" class="codepen">Subukan sa Pen ang  <a href="https://codepen.io/gaearon/pen/BwWzwm">Paano mag-isip sa React: Ika-2 Hakbang</a> sa <a href="https://codepen.io">CodePen</a>.</p>
<script async src="https://production-assets.codepen.io/assets/embed/ei.js"></script>

Ngayon na mayroon ka nang component hierarchy, oras na para gawin ito sa iyong app. Ang pinakamadaling paraan ay bumuo ng version na tatanggap ng data model at irerender sa UI ngunit wala itong interactivity. Mas makakabuti na paghiwalayin ang mga prosesong ito dahil ang pagbuo ng static version ay nangangailangan ng maraming typing at kaunting pag-iisip, at ang pag-dagdag naman ng interactivity ay nangangailangan ng malalim na pag-iisip at kaunting pagta-type. Malalaman nating kung bakit.

> Upang makabuo ka ng static version ng app na magrerender ng iyong data model, gugustuhin mong bumuo muna ng components na maaaring gamitin pa ng iba pang components at magpasa ng data gamit ang *props*. Ang *props* ay paraan ng pagpasa ng data mula sa parent papunta sa child component. Kung ikaw ay pamilyar sa concept ng *state*, **huwag lang gumamit ng state** upang bumuo ng static version. Ang State ay nakareserba lamang para sa interactivity, yan ay, sa mga data nagbabago sa paglipas ng panahon.

Reviewer: ```suggestion
Upang makabuo ka ng static version ng app na magrerender ng iyong data model, gugustuhin mong bumuo muna ng components na maaaring gamitin pa ng iba pang components at magpasa ng data gamit ang *props*. Ang *props* ay paraan ng pagpasa ng data mula sa parent papunta sa child component. Kung ikaw ay pamilyar sa concept ng *state*, **huwag kang gumamit ng state** upang bumuo ng static version. Ang State ay nakareserba lamang para sa interactivity, yan ay, sa mga data na nagbabago o *dynamic*.
```
